[doc] Add doc for importing SavedModel from TF Hub (#5523)
google/iree: mkdocs/docs/frontends/tensorflow.md

> ## Importing models
> 
> ### From SavedModel on TensorFlow Hub
> 
> IREE supports importing and using SavedModels from
> [TensorFlow Hub](https://www.tensorflow.org/hub).
> 
> #### Using the command-line tool
> 
> IREE compilers transform a model into its final deployable format in many
> sequential steps. The first step is to import the model authored with Python
> in TensorFlow into a format (i.e., [MLIR](https://mlir.llvm.org/)) expected
> by main IREE compilers. This can be done via the `iree-tf-import` tool.

Reviewer: Let's shuffle this paragraph around a bit. The general steps apply to SavedModels not downloaded from TF Hub _and_ to using the Python compiler APIs (which invoke the command line tool(s)).

Also included in my suggestion:

* "many sequential steps" ->"several sequential steps": rewording to be less daunting to users ;)
* "expected by main IREE compilers" -> "compatible with the generic IREE compilers": not sold on either language, but I want to hint more strongly that each framework importer acts in a similar way - converging on common formats

```suggestion
## Importing models

IREE compilers transform a model into its final deployable format in several
sequential steps. The first step for a TensorFlow model is to use either the
`iree-tf-import` command-line tool or IREE's Python APIs to import the model
into a format (i.e., [MLIR](https://mlir.llvm.org/)) compatible with the generic
IREE compilers.

### From SavedModel on TensorFlow Hub

IREE supports importing and using SavedModels from
[TensorFlow Hub](https://www.tensorflow.org/hub).

#### Using the command-line tool
```

Author: Cool. I'll just blatantly copy your words here. :)

sequential steps. The first step is to import the model authored with Python
in TensorFlow into a format (i.e., [MLIR](https://mlir.llvm.org/)) expected
by main IREE compilers. This can be done via the `iree-tf-import` tool.

!!! note
    `iree-tf-import` is installed as `/path/to/python/site-packages/iree/tools/tf/iree-tf-import`
    via the `iree-tools-tf-snapshot` Python package. You can find out the full
    path to the `site-packages` directory via the `python -m site` command.

First download the SavedModel and re-save it with a [serving signature](https://www.tensorflow.org/guide/saved_model#specifying_signatures_during_export).

!!! info
    This is necessary because by default TensorFlow Hub SavedModels might not

>     have serving signatures, which are required for IREE compilation flow.

Reviewer: To expand on this a bit further:

For cases that most users will hit, we should strive for the tools to do the heavy lifting. If that means editing `iree-tf-import` or writing a separate tool to re-save with serving signatures (in this case), we should prefer doing that over adding more steps to our documentation.

For cases that are rarer, or optional (such as for certain optimizations), we can defer to a "troubleshooting" or "tuning performance" section.

>     have serving signatures, which are required for IREE compilation flow.

Reviewer: (and it's great that you spotted this and figured out a way to make it work, doc writing and going through user flows is expected to uncover areas to improve like this, thanks!)

>     have serving signatures, which are required for IREE compilation flow.

Author: Agreed with your points. I've actually chatted with @silvasean and a mode to inspect possible entry points / exported names could be very useful and it could reduce the number of steps we require the user to take here. However, I'm not sure that we can go as far as to use `iree-tf-import` to modify the saved model. But @silvasean knows more there.

I agree with you that we should move the resaving into an optional step. I'm using MobileNet as an example here and I'd suspect it applies to other models from TF Hub, but indeed  some SavedModel might already have that. So restructured a bit. Let me know what you think now.

>     have serving signatures, which are required for IREE compilation flow.

Reviewer 2: We could try a few more to see if there are shortcuts to take in practice, but unfortunately due to how SavedModel's work, the only really reliable way to know what you're getting is to load them in Python and save them in a controlled fashion.

SignatureDef saved models (i.e. "v1", though that's a misnomer as they are produced by both TF1 and TF2) are somewhat easier to handle. We can list the serving signatures, and if there is sufficiently good shape information in the model already then the user can just choose. For ObjectGraph saved models (i.e. "v2"), it gets complicated.

!!! note

>     If there is no serving signatures in the original SavedModel, you may check
>     to add it by yourself by following ["Missing serving signature in
>     SavedModel"](#missing-serving-signature-in-savedmodel).

Reviewer: grammar (and some formatting, but the grammar is what matters)

```suggestion
    If there are no serving signatures in the original SavedModel, you may add
    them yourself by following
    ["Missing serving signature in SavedModel"](#missing-serving-signature-in-savedmodel).
```

End-to-end execution tests can be found in IREE's
[integrations/tensorflow/e2e/](https://github.com/google/iree/tree/main/integrations/tensorflow/e2e)
directory.

## Troubleshooting

### Missing serving signature in SavedModel

> Sometimes SavedModel are exported without explicit [serving signatures](https://www.tensorflow.org/guide/saved_model#specifying_signatures_during_export).

Reviewer: nit: grammar (plural)

```suggestion
Sometimes SavedModels are exported without explicit [serving signatures](https://www.tensorflow.org/guide/saved_model#specifying_signatures_during_export).
```
